### PR TITLE
Vacancy templates

### PR DESCRIPTION
**Pull request by muhtimur**
# Contribution checklist

## Brief description

Created the same as Issue templates but for vacancies.
- Description
- Short description
- Allow to set custom fields
- Allow to set a set of related issues

image

## Checklist

* [ ] - UI test added to added/changed functionality?
* [x] - Screenshot is added to PR if applicable ?
* [x] - Does the code work? Check whether function and logic are correct.
* [ ] - Does Changelog.md is updated with changes?
* [x] - Does the translations are up to date?
* [x] - Does it well tested?
* [x] - Tested for Chrome.
* [x] - Tested for Safari.
* [x] - Go through the changed code looking for typos, TODOs, commented LOCs, debugging pieces of code, etc.
* [x] - Rebase your branch onto master and upstream branch
* [x] - Is there any redundant or duplicate code?
* [ ] - Are required links are linked to PR?
* [ ] - Does new code is well documented ?

## Related issues

[TSK-337](https://front.hc.engineering/workbench/platform/setting/setting/classes#tracker%3Acomponent%3AEditIssue%7C634d081a6ae32428a9cefb9d%7Ctracker%3Aclass%3AIssue%7Ccontent)

### CODE REVIEW

File: plugins/task-resources/src/components/state/StatesEditor.svelte [hunk 1]
```diff
@@ -83,8 +95,76 @@
...
+          content={template?.description ?? ''}
+          on:value={(evt) => onDescriptionChange(evt.detail)}
+        />
+      {/key}
+    </div>
+  </div>
+  <div class="antiSection mt-9 mb-9">
+    <div class="antiSection-header">
+      <div class="antiSection-header__icon">
+        <Icon icon={tracker.icon.Issue} size={'small'} />
```
Comment (BykhovDenis) [inline plugins/task-resources/src/components/state/StatesEditor.svelte]: Better move this icon to task and remove tracker dependency

File: plugins/task-resources/src/components/state/StatesEditor.svelte [hunk 2]
```diff
@@ -15,25 +15,34 @@
...
   import { ColorsPopup } from '@hcengineering/view-resources'
+  import { StyledTextBox } from '@hcengineering/text-editor'
+  import tracker from '@hcengineering/tracker'
+  import { getFiltredKeys } from '@hcengineering/view-resources/src/utils'
   import Circles from './Circles.svelte'
   import StatusesPopup from './StatusesPopup.svelte'
   import task from '../../plugin'
   import Won from '../icons/Won.svelte'
   import Lost from '../icons/Lost.svelte'
+  import recruit from '../../../../recruit-resources/src/plugin'
```
Comment (BykhovDenis) [inline plugins/task-resources/src/components/state/StatesEditor.svelte]: Please remove it and move `RelatedIssues` string to task plugin

File: plugins/task/src/index.ts [hunk 3]
```diff
@@ -167,6 +167,8 @@ export interface LostStateTemplate extends DoneStateTemplate, LostState {}
  */
 export interface KanbanTemplate extends Doc {
   title: string
+  description: string
```
Comment (BykhovDenis) [inline plugins/task/src/index.ts]: Probably we can do it optional?

**Review by BykhovDenis — CHANGES_REQUESTED** [review]

File: plugins/task-resources/src/components/state/StatesEditor.svelte (hunk 1 above)
Comment (muhtimur) [reply plugins/task-resources/src/components/state/StatesEditor.svelte]: So what about components from tracker `tracker.component.CreateIssue` and `tracker.component.RelatedIssues`? I use them in this module too

File: plugins/task-resources/src/components/state/StatesEditor.svelte (hunk 2 above)
Comment (muhtimur) [reply plugins/task-resources/src/components/state/StatesEditor.svelte]: I also use `recruit.class.Vacancy` from recruit package in this module. I think this package should be added to task-resources. Isnt it?

File: plugins/task/src/index.ts (hunk 3 above)
Comment (muhtimur) [reply plugins/task/src/index.ts]: Done

File: plugins/task-resources/src/components/state/StatesEditor.svelte (hunk 2 above)
Comment (BykhovDenis) [reply plugins/task-resources/src/components/state/StatesEditor.svelte]: I think we should create some optional components contributing in models as field or mixing to KanbanTemplateSpace. Also I think we need create new related issue for each vacancy.

File: models/tracker/src/index.ts [hunk 4]
```diff
@@ -302,6 +302,9 @@ export class TIssueTemplate extends TDoc implements IssueTemplate {
 
   @Prop(Collection(attachment.class.Attachment), tracker.string.Attachments)
     attachments!: number
+
+  @Prop(ArrOf(TypeRef(core.class.TypeRelatedDocument)), tracker.string.RelatedTo)
```
Comment (haiodo) [inline models/tracker/src/index.ts:306]: Why we need them here?

File: plugins/recruit-resources/src/components/CreateVacancy.svelte [hunk 5]
```diff
@@ -143,6 +179,19 @@
       }}
     />
   </svelte:fragment>
+  {#key template?.description}
```
Comment (haiodo) [inline plugins/recruit-resources/src/components/CreateVacancy.svelte:182]: We need description in any case, and it by default should be taken from template?.description, but user should be able to create vacancy without template with description.

File: plugins/recruit-resources/src/components/CreateVacancy.svelte [hunk 6]
```diff
@@ -117,20 +167,6 @@
...
-    _class={recruit.class.Vacancy}
-    space={objectId}
-    alwaysEdit
-    showButtons={false}
-    maxHeight={'card'}
-    bind:content={fullDescription}
-    placeholder={recruit.string.FullDescription}
-  />
-  <svelte:fragment slot="pool">
     <Component
```
Comment (haiodo) [inline plugins/recruit-resources/src/components/CreateVacancy.svelte:170]: I think we need to move vacancy template to header, same way we have for Tracker issues.

File: plugins/task-resources/src/components/state/StatesEditor.svelte [hunk 7]
```diff
@@ -83,8 +94,77 @@
...
+  async function onDescriptionChange (value: string) {
+    dispatch('descriptionChange', { value })
+  }
+
+  async function onShortDescriptionChange (value: string) {
+    dispatch('shortDescriptionChange', { value })
+  }
 </script>
 
+{#if template?.space === 'recruit:space:VacancyTemplates' && space}
```
Comment (haiodo) [inline plugins/task-resources/src/components/state/StatesEditor.svelte]: Please make it configurable, so you need some mixin added to space or template, so task-resources should do not know about recruit.

File: plugins/task/src/index.ts [hunk 8]
```diff
@@ -178,6 +180,8 @@ export interface KanbanTemplateSpace extends Doc {
   name: IntlString
   description: IntlString
   icon: AnyComponent
+  relatedIssueTemplatesComponent?: AnyComponent
```
Comment (haiodo) [inline plugins/task/src/index.ts]: This should be configurable via Mixin, so any kind of template configuration will be available.

File: plugins/tracker-resources/src/components/issues/RelationEditorPart.svelte [hunk 9]
```diff
@@ -91,10 +93,15 @@
     <div class="tag-container">
       <div class="flex-grow">
         <div class="overflow-label max-w-30">
-          <Component
-            is={view.component.ObjectPresenter}
-            props={{ objectId: doc._id, _class: doc._class, value: doc, inline: true }}
-          />
+          {#if doc._class === 'task:class:KanbanTemplate'}
```
Comment (haiodo) [inline plugins/tracker-resources/src/components/issues/RelationEditorPart.svelte]: Please use 
```
import task, { KanbanTemplate } from '@hcengineering/task'
...
{#if doc._class === task.class.KanbanTemplate}
... 
```

Comment (haiodo) [reply plugins/tracker-resources/src/components/issues/RelationEditorPart.svelte]: Also, why we need to show it here? Relation's will be shown for real issues, so I suppose this need to be removed.

File: plugins/tracker/src/index.ts [hunk 10]
```diff
@@ -258,6 +258,8 @@ export interface IssueTemplate extends Doc, IssueTemplateData {
   // Discussion stuff
   comments: number
   attachments?: number
+
+  relations?: RelatedDocument[]
```
Comment (haiodo) [inline plugins/tracker/src/index.ts:262]: Why we need relations for template? I supposed we will have IssueTemplate instances per Vacancy template, it we will add normal relations for created issues.

**Review by haiodo — CHANGES_REQUESTED** [review]
Some changes are required.

File: plugins/tracker-resources/src/components/issues/RelationEditorPart.svelte (hunk 9 above)
Comment (muhtimur) [reply plugins/tracker-resources/src/components/issues/RelationEditorPart.svelte]: Done

File: plugins/recruit-resources/src/components/CreateVacancy.svelte (hunk 6 above)
Comment (muhtimur) [reply plugins/recruit-resources/src/components/CreateVacancy.svelte:170]: I moved it to the 'Create Vacancy' dialog header like in the tracker:

image
image


Did you mean something else?

Comment (haiodo) [reply plugins/recruit-resources/src/components/CreateVacancy.svelte:170]: Nope, this is fine.

File: plugins/task-resources/src/components/state/StatesEditor.svelte (hunk 7 above)
Comment (muhtimur) [reply plugins/task-resources/src/components/state/StatesEditor.svelte]: Done

File: plugins/task/src/index.ts (hunk 8 above)
Comment (muhtimur) [reply plugins/task/src/index.ts]: Done

File: plugins/tracker/src/index.ts (hunk 10 above)
Comment (muhtimur) [reply plugins/tracker/src/index.ts:262]: If we want to create 'IssueTemplates' field in Vacancy template (KanbanTemplate interface), we should add 'tracker' dependency, but this solution doesnt need it

**Review by haiodo — APPROVED** [review]

**Review by BykhovDenis — APPROVED** [review]

**Pull request merged**